app/src/main/java/ca/ualberta/CMPUT3012019T02/alexandria/controller/BookController.java

```diff
@@ -485,25 +469,65 @@ public void onCancelled(@NonNull DatabaseError databaseError) {
     }
 
-    private <T extends UserBook> CompletableFuture<Void> updateUserBook(@NonNull UserBookType<T> userBookType, @NonNull String id, @NonNull T userBook) {
-        return updateUserBook(userBookType, id, userBook.getIsbn(), userBook);
+    private <T extends UserBook> CompletableFuture<Void> updateUserBook(@NonNull String userBookPath, @NonNull String id, @NonNull T userBook) {
+        return updateUserBook(userBookPath, id, userBook.getIsbn(), userBook);
     }
 
-    private <T extends UserBook> CompletableFuture<Void> deleteUserBook(@NonNull UserBookType<T> userBookType, @NonNull String id, @NonNull String isbn) {
-        return updateUserBook(userBookType, id, isbn, null);
+    private <T extends UserBook> CompletableFuture<Void> deleteUserBook(@NonNull String userBookPath, @NonNull String id, @NonNull String isbn) {
+        return updateUserBook(userBookPath, id, isbn, null);
     }
 
-    private <T extends UserBook> CompletableFuture<Void> updateUserBook(@NonNull UserBookType<T> userBookType, @NonNull String id, @NonNull String isbn, T userBook) {
+    private <T extends UserBook> CompletableFuture<Void> updateUserBook(@NonNull String userBookPath, @NonNull String id, @NonNull String isbn, T userBook) {
         final CompletableFuture<Void> future = new CompletableFuture<>();
 
-        getUserBookReference(userBookType, id, isbn).setValue(userBook)
+        getUserBookReference(userBookPath, id, isbn).setValue(userBook)
                 .addOnSuccessListener(future::complete)
                 .addOnFailureListener(future::completeExceptionally);
 
         return future;
     }
 
 
+    /* Updating Available Book Owners */
+
+
+    private CompletableFuture<Void> addAvailableOwner(String id, String isbn) {
+        return CompletableFuture.runAsync(() -> {
+            try {
+                Book book = getBook(isbn).get(5, TimeUnit.SECONDS);
+                if (book.getAvailableOwners().contains(id)) {
+                    return;
+                }
+                book.addAvailableOwners(id);
+                updateBook(book).get(5, TimeUnit.SECONDS);
+            } catch (InterruptedException e) {
+                e.printStackTrace();
+            } catch (ExecutionException e) {
+                e.printStackTrace();
+            } catch (TimeoutException e) {
+                e.printStackTrace();
+            }
+        });
+    }
+
+    private CompletableFuture<Void> removeAvailableOwner(String id, String isbn) {
+        return CompletableFuture.supplyAsync(() -> {
+            try {
+                Book book = getBook(isbn).get(5, TimeUnit.SECONDS);
+                book.removeAvailableOwners(id);
+                updateBook(book).get(5, TimeUnit.SECONDS);
+            } catch (InterruptedException e) {
+                e.printStackTrace();
+            } catch (ExecutionException e) {
+                e.printStackTrace();
+            } catch (TimeoutException e) {
+                e.printStackTrace();

```

Comment:
Same here. I am not sure if there is a problem with " getBook(isbn).get(5, TimeUnit.SECONDS);" how it can return an exception